Fix ruby 3.4 compatibility

PikachuEXE opened the pull request:
Same as https://github.com/pganalyze/pg_query/pull/320 but that PR got no workflow to approve anymore and author not responding for 5 days

msakrejda commented:
Thanks for opening a new PR!

PikachuEXE commented:
No idea what's wrong

seanlinsley commented:
I'm not sure what changed in Ruby 3.4 for this require to no longer be needed, but we'll surely need to still require it for older Ruby versions. For example:

```rb
require 'pg_query/pg_query' unless RUBY_VERSION > '3.4'
```

Also note that you'll want to add Ruby 3.4 to the [CI matrix](https://github.com/pganalyze/pg_query/blob/main/.github/workflows/ci.yml#L11). But with the comment on that line, there may be some extra work to support Ruby 3.4 in the protobuf gem.

PikachuEXE commented:
Updated, added 3.4.0-rc1 to matrix (would `3.4` pick up rc versions? I dunno

Found `[google-protobuf](https://rubygems.org/gems/google-protobuf)` PR for ruby 3.4 https://github.com/protocolbuffers/protobuf/pull/19752 but CI failing so no idea if it supports 3.4 yet

PikachuEXE commented:
I just run locally to see without workflow approval

Without the require some methods are not defined
![image](https://github.com/user-attachments/assets/a17613b1-a2dc-4f88-aede-85c981c6a6fc)

Adding back the `require` brings us back to the same issue...
![image](https://github.com/user-attachments/assets/5d65780d-2ee8-4e86-a45b-0d731126db72)

PikachuEXE commented:
Turns out it might be development version specific issue
See https://github.com/grpc/grpc/pull/31970 (I took some code changes from that gem's latest version)

Locally run on 3.2, 3.3, 3.4.0-rc1
![image](https://github.com/user-attachments/assets/2a88f2ac-b59b-48f9-9f96-5d6ab6376011)

No idea how to test freebsd one (not in CI matrix as well)

PikachuEXE commented:
Update `Gemfile.lock` I guess...?
(For my own gems I never got it checked-in

Also `3.4.0-rc1` is unavailable on windows. Do you prefer skipping it or using head?
![2024-12-21 04_13_19-Window](https://github.com/user-attachments/assets/1d48f862-5078-4034-8857-5adf742fa3ea)

PikachuEXE commented:
Yes I forgot to run bundle update, my fault...

msakrejda commented:
It looks like it's still failing on 3.4 due to protobuf issues, but it's also failing on 2.7. I think for 2.7, we could consider dropping it, since that's been out of support for over a year...

PikachuEXE commented:
Ya `google-protobuf` native gems are declared to be incompatible with 3.4 until 3.4 released
We can still test 2.7 if `Gemfile.lock` is dropped.

PikachuEXE commented:
Do we want to try using `force_ruby_platform` just to see if CI passes? (Can still wait for native version to be released

PikachuEXE commented:
New version of `google-protobuf` released supporting ruby 3.4 - https://rubygems.org/gems/google-protobuf/versions/4.29.3-x86_64-linux

This PR is updated to point to 3.4 (also removed 2.7 like you said

lfittl reviewed (approved):
:+1: Thanks for the contribution & your patience on getting this merged. FWIW, I did make a few edits to the extconf.rb code to make it shorter and easier to understand.

chaadow commented:
hey @lfittl is it possible to cut a release please 🙏🏼 thanks

chaadow commented:
there is also an incompatibility with ruby 4.0.0-preview2, if anyone got time to prepare for the release on december 25th, that would be great 🙏🏼 ( i have no clue what's happening otherwise i would have contributed ) 

https://github.com/pganalyze/pg_query/issues/340
